File: SpaceInvaders/Timer/TimeEvent.cs
using System;
using System.Diagnostics;

namespace SpaceInvaders
{
    public class TimeEvent : DLink
    {
        private Name name;
        private Command pCommand;
        private float triggerTime;
        private float deltaTime;

        public TimeEvent()
            : base()
        {
            this.name = TimeEvent.Name.Uninitialized;
            this.pCommand = null;
            this.triggerTime = 0.0f;
            this.deltaTime = 0.0f;
        }

        ~TimeEvent()
        {
            this.name = Name.Uninitialized;
            this.pCommand = null;
        }

        public void Set(TimeEvent.Name eventName, Command pCommand, float deltaTimeToTrigger)
        {
            Debug.Assert(pCommand != null);

            this.name = eventName;
            this.pCommand = pCommand;
            this.deltaTime = deltaTimeToTrigger;

            // set the trigger time
            this.triggerTime = TimerManager.GetCurrTime() + deltaTimeToTrigger;
        }

        public void Process()
        {
            Debug.Assert(this.pCommand != null);
            this.pCommand.Execute(deltaTime);
        }

        public override void Clean()
        {
            base.Clean();
            this.name = TimeEvent.Name.Uninitialized;
            this.pCommand = null;
            this.triggerTime = 0.0f;
            this.deltaTime = 0.0f;
        }

        public override void Print()
        {
            Debug.WriteLine("   Name: {0} ({1})", this.name, this.GetHashCode());
            
            Debug.WriteLine("      Command: {0}", this.pCommand);
            Debug.WriteLine("   Event Name: {0}", this.name);
            Debug.WriteLine(" Trigger Time: {0}", this.triggerTime);
            Debug.WriteLine("   Delta Time: {0}", this.deltaTime);

            // TODO Consolidate like below
            //string next = (this.pNext != null) ? this.pNext : second_expression;

            /*
            if (this.pNext == null)
            {
                Debug.WriteLine("      next: null");
            }
            else
            {
                Image pTmp = (Image)this.pNext;
                Debug.WriteLine("      next: {0} ({1})", pTmp.name, pTmp.GetHashCode());
            }

            if (this.pPrev == null)
            {
                Debug.WriteLine("      prev: null");
            }
            else
            {
                Image pTmp = (Image)this.pPrev;
                Debug.WriteLine("      prev: {0} ({1})", pTmp.name, pTmp.GetHashCode());
            }
            */

            base.Print();
        }

        public void SetName(Name newName)
        {
            this.name = newName;
        }

        public Name GetName()
        {
            return name;
        }

        public Command GetCommand()
        {
            return this.pCommand;
        }

        public float GetDeltaTime()
        {
            return this.deltaTime;
        }

        public void SetTriggerTime(float triggerTime)
        {
            this.triggerTime = triggerTime;
        }

        public float GetTriggerTime()
        {
            return triggerTime;
        }
        
        public void SetCommand(Command newCommand)
        {
            this.pCommand = newCommand;
        }

        public enum Name
        {
            SpriteAnimation,
            SpriteMovement,
            MissileMovement,
            AlienGridMovementSound,
            DropBomb,

            DebugCommand,
            Uninitialized,
            DeployFlyingSaucer,
            ResetLevel,
            AlienGridMovement,
            ResetCoreCannon,
        }
    }
}

File: quarkus-migration/space-invaders-quarkus/src/main/java/net/beeland/spaceinvaders/timer/TimeEvent.java
package net.beeland.spaceinvaders.timer;

import net.beeland.spaceinvaders.command.Command;
import net.beeland.spaceinvaders.manager.DLink;

/**
 * TimeEvent represents a scheduled command execution.
 * Events are stored in a sorted list by trigger time and executed
 * when the current time reaches or exceeds the trigger time.
 */
public class TimeEvent extends DLink {
    
    /**
     * Event name enumeration for identifying different types of timer events
     */
    public enum Name {
        SPRITE_ANIMATION,
        SPRITE_MOVEMENT,
        MISSILE_MOVEMENT,
        ALIEN_GRID_MOVEMENT_SOUND,
        DROP_BOMB,
        DEBUG_COMMAND,
        DEPLOY_FLYING_SAUCER,
        RESET_LEVEL,
        ALIEN_GRID_MOVEMENT,
        RESET_CORE_CANNON,
        ANIMATION,
        UNINITIALIZED
    }
    
    private Name name;
    private Command command;
    private float triggerTime;
    private float deltaTime;
    
    /**
     * Default constructor
     */
    public TimeEvent() {
        super();
        this.name = Name.UNINITIALIZED;
        this.command = null;
        this.triggerTime = 0.0f;
        this.deltaTime = 0.0f;
    }
    
    /**
     * Set the time event properties
     * 
     * @param eventName The name/type of this event
     * @param command The command to execute when triggered
     * @param deltaTimeToTrigger Time from now until trigger
     * @param currentTime The current game time
     */
    public void set(Name eventName, Command command, float deltaTimeToTrigger, float currentTime) {
        assert command != null : "Command cannot be null";
        assert deltaTimeToTrigger >= 0.0f : "Delta time must be non-negative";
        
        this.name = eventName;
        this.command = command;
        this.deltaTime = deltaTimeToTrigger;
        this.triggerTime = currentTime + deltaTimeToTrigger;
    }
    
    /**
     * Process/execute this time event
     */
    public void process() {
        assert this.command != null : "Cannot process event with null command";
        this.command.execute(this.deltaTime);
    }
    
    @Override
    protected void wash() {
        this.clear();
        this.name = Name.UNINITIALIZED;
        this.command = null;
        this.triggerTime = 0.0f;
        this.deltaTime = 0.0f;
    }
    
    @Override
    public void dump() {
        System.out.println("TimeEvent:");
        System.out.println("  Name: " + this.name);
        System.out.println("  Command: " + (this.command != null ? this.command.getClass().getSimpleName() : "null"));
        System.out.println("  Trigger Time: " + this.triggerTime);
        System.out.println("  Delta Time: " + this.deltaTime);
    }
    
    // Getters and Setters
    
    public Name getName() {
        return this.name;
    }
    
    public void setName(Name name) {
        this.name = name;
    }
    
    public Command getCommand() {
        return this.command;
    }
    
    public void setCommand(Command command) {
        this.command = command;
    }
    
    public float getTriggerTime() {
        return this.triggerTime;
    }
    
    public void setTriggerTime(float triggerTime) {
        this.triggerTime = triggerTime;
    }
    
    public float getDeltaTime() {
        return this.deltaTime;
    }
    
    public void setDeltaTime(float deltaTime) {
        this.deltaTime = deltaTime;
    }
}